homeassistant/components/improv_ble/config_flow.py

```diff
@@ -0,0 +1,368 @@
+                discovery.address in current_addresses
+                or discovery.address in self._discovered_devices
+                or not device_filter(discovery.advertisement)
+            ):
+                continue
+            self._discovered_devices[discovery.address] = discovery
+
+        if not self._discovered_devices:
+            return self.async_abort(reason="no_devices_found")
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_ADDRESS): vol.In(
+                    {
+                        service_info.address: (
+                            f"{service_info.name} ({service_info.address})"
+                        )
+                        for service_info in self._discovered_devices.values()
+                    }
+                ),
+            }
+        )
+        return self.async_show_form(
+            step_id="user",
+            data_schema=data_schema,
+            errors=errors,
+        )
+
+    async def async_step_bluetooth(
+        self, discovery_info: BluetoothServiceInfoBleak
+    ) -> FlowResult:
+        """Handle the Bluetooth discovery step."""
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        service_data = discovery_info.service_data
+        improv_service_data = ImprovServiceData.from_bytes(
+            service_data[SERVICE_DATA_UUID]
+        )
+        if improv_service_data.state in (State.PROVISIONING, State.PROVISIONED):
+            _LOGGER.debug(
+                "Device is already provisioned: %s", improv_service_data.state
+            )
+            return self.async_abort(reason="already_provisioned")
+        self._discovery_info = discovery_info
+        name = self._discovery_info.name or self._discovery_info.address
+        self.context["title_placeholders"] = {"name": name}
+        return await self.async_step_bluetooth_confirm()
+
+    async def async_step_bluetooth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle bluetooth confirm step."""
+        # mypy is not aware that we can't get here without having these set already
+        assert self._discovery_info is not None
+
+        if user_input is None:
+            name = self._discovery_info.name or self._discovery_info.address
+            return self.async_show_form(
+                step_id="bluetooth_confirm",
+                description_placeholders={"name": name},
+            )
+
+        return await self.async_step_start_improv()
+
+    async def async_step_start_improv(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Start improv flow.
+
+        If the device supports identification, show a menu, if it does not,
+        ask for WiFi credentials.
+        """
+        # mypy is not aware that we can't get here without having these set already
+        assert self._discovery_info is not None
```

Comment:
We should fetch the latest service info again here and see if its already provisioned and than abort if it is since it may have been provisioned by another HA instance. https://developers.home-assistant.io/docs/core/bluetooth/api?_highlight=bluetooth#fetching-the-latest-bluetoothserviceinfobleak-for-a-device

Reply:
I don't agree with this, it can't be the responsability of every BLE integration to keep discovery data up to date.
The bluetooth integration should abort flows and re-trigger them when advertised data has changed.

Reply:
To give some more context. Most of the adv changes come from:

- Sensors that report their state via adv.  some even report temperature up to 4 decimal places and very frequently 

- devices that report their current state. Many devices including switch bot have the state of the curtain etc in the advertisement (ie percentage open)


- finally since advertising is limited to 31 bytes, devices cycle through advertisements to communicate data so the last advertisement may change 5 times quickly to send a burst of data

Reply:
The flow now checks for latest discovery data before provisioning, but I don't really like it TBH. I can imagine two ways to avoid it:
1. An integration with bluetooth discovery is allowed to register a "smarter" check than what's possible via the filter in `manifest.json`, this filter would then be re-run on every BLE advertisement change. If the filter returns `True`, a flow is started, if it returns `False`, an already started flow is aborted.
2. Let bluetooth discovery flows subscribe to BLE advertisement changes via `bluetooth.async_register_callback`, and abort a flow as needed. The drawback of this is that it won't trigger rediscovery of factory reset `improv_ble` device